src/library.js

```diff
@@ -3241,13 +3241,15 @@ LibraryManager.library = {
     buf >>= 2;
     while (ch = HEAPU8[sigPtr++]) {
 #if ASSERTIONS
-      assert(ch === 100/*'d'*/ || ch === 102/*'f'*/ || ch === 105 /*'i'*/);
+      assert(ch === 100/*'d'*/ || ch === 102/*'f'*/ || ch === 105 /*'i'*/ || ch === 106 /*'j'*/);
 #endif
       // A double takes two 32-bit slots, and must also be aligned - the backend
       // will emit padding to avoid that.
       var double = ch < 105;
       if (double && (buf & 1)) buf++;
-      readAsmConstArgsArray.push(double ? HEAPF64[buf++ >> 1] : HEAP32[buf]);
+      var arg = double ? HEAPF64[buf++ >> 1] : HEAP32[buf];
+      if (ch === 106) arg = JsValStore.get(arg);
+      readAsmConstArgsArray.push(arg);
```

Comment:
This is nice idea, but we have not actually taken the stance that EM_ASM, EM_JS or calling JS library functions would automatically marshal high level objects across the boundary, e.g. JS strings and arrays are not automatically converted either.

(Also, `j` is already used to represent `int64`, so would need to use some other character)

This PR would increase generated code size for all users, even those that don't use WebGPU and don't need such marshalling. I.e. the `JsValStore` object would now become an un-dceable part of the core runtime that anyone who has a EM_ASM or a EM_JS reference in their codebase would then pull in.

(Also using EM_ASM except for special circumstances, simple unit tests and prototyping is a bad practice and good to be avoided. The issue is that it unfortunately generates poor code compared to using proper JS libraries, which are the preferred method)

I would recommend handling the object wrapping at the caller before returning, that will avoid building up boilerplate code.

Reply:
@juj Does what you say there apply to EM_JS too? Or does EM_ASM produce particular poor code?

Reply:
See #12778. After #13443 EM_JS should generate better code, but have not had a chance to verify.

Reply:
Ok, I updated this so that EM_ASM does not automatically pull in the marshalling. You said that EM_ASM shouldn't be used except for special circumstances. Is there a way to do this without EM_ASM?